File: agents/Group37/mcts.py
import math
from copy import deepcopy

from src.Board import Board
from src.Colour import Colour
from src.Tile import Tile


class MCTSNode:
    """Node in the MCTS tree."""

    def __init__(
        self,
        board: Board,
        player_to_move: Colour,
        parent: "MCTSNode | None" = None,
        move: tuple[int, int] | None = None,
    ):
        self.board = board
        self.player_to_move = player_to_move
        self.parent = parent
        self.move = move  # Move that led to this node
        self.children: dict[tuple[int, int], "MCTSNode"] = {}
        self.visits = 0
        self.wins = 0.0
        self.untried_moves = self._get_empty_tiles()

    def _get_empty_tiles(self) -> list[tuple[int, int]]:
        """Get all empty positions on the board."""
        empty = []
        for i in range(self.board.size):
            for j in range(self.board.size):
                if self.board.tiles[i][j].colour is None:
                    empty.append((i, j))
        return empty

    def is_fully_expanded(self) -> bool:
        return len(self.untried_moves) == 0

    def is_terminal(self) -> bool:
        """Check if this is a terminal state (someone won or board full)."""
        return self._check_winner() is not None or len(self._get_empty_tiles()) == 0

    def _check_winner(self) -> Colour | None:
        """Check if there's a winner. Returns the winning color or None."""
        # Check RED (top to bottom)
        if self._has_won(Colour.RED):
            return Colour.RED
        # Check BLUE (left to right)
        if self._has_won(Colour.BLUE):
            return Colour.BLUE
        return None

    def _has_won(self, colour: Colour) -> bool:
        """Check if the given color has won using DFS."""
        size = self.board.size
        visited = set()

        if colour == Colour.RED:
            # Start from top row, try to reach bottom row
            start_positions = [
                (0, j)
                for j in range(size)
                if self.board.tiles[0][j].colour == colour
            ]
            target_row = size - 1

            def is_goal(x, y):
                return x == target_row

        else:  # BLUE
            # Start from left column, try to reach right column
            start_positions = [
                (i, 0)
                for i in range(size)
                if self.board.tiles[i][0].colour == colour
            ]
            target_col = size - 1

            def is_goal(x, y):
                return y == target_col

        # DFS from each starting position
        for start in start_positions:
            if start in visited:
                continue
            stack = [start]
            while stack:
                x, y = stack.pop()
                if (x, y) in visited:
                    continue
                visited.add((x, y))
                if is_goal(x, y):
                    return True
                # Check neighbors
                for k in range(Tile.NEIGHBOUR_COUNT):
                    nx = x + Tile.I_DISPLACEMENTS[k]
                    ny = y + Tile.J_DISPLACEMENTS[k]
                    if 0 <= nx < size and 0 <= ny < size:
                        if (nx, ny) not in visited:
                            if self.board.tiles[nx][ny].colour == colour:
                                stack.append((nx, ny))
        return False

    def ucb1(self, exploration_constant: float = 1.414) -> float:
        """Calculate UCB1 value for this node."""
        if self.visits == 0:
            return float("inf")
        exploitation = self.wins / self.visits
        exploration = exploration_constant * math.sqrt(
            math.log(self.parent.visits) / self.visits
        )
        return exploitation + exploration

    def best_child(self, exploration_constant: float = 1.414) -> "MCTSNode":
        """Select the best child according to UCB1."""
        return max(
            self.children.values(),
            key=lambda c: c.ucb1(exploration_constant),
        )

    def expand(self) -> "MCTSNode":
        """Expand by adding a new child node for an untried move."""
        move = self.untried_moves.pop()
        new_board = deepcopy(self.board)
        new_board.set_tile_colour(move[0], move[1], self.player_to_move)
        next_player = Colour.opposite(self.player_to_move)
        child = MCTSNode(new_board, next_player, parent=self, move=move)
        self.children[move] = child
        return child

File: agents/Group37/cpp/mcts.cpp
#include "mcts.h"
#include "heuristics.h"
#include "transposition_table.h"
#include <chrono>
#include <random>
#include <algorithm>
#include <limits>
#include <iostream>

namespace hex {

std::pair<int, int> find_winning_move(const Board& board, Colour colour) {
    auto empty_tiles = board.get_empty_tiles();

    // Only check if few empty tiles (more likely to have winning move)
    if (empty_tiles.size() > 30) {
        return {-1, -1};
    }

    for (const auto& move : empty_tiles) {
        // Make temporary move
        Board temp_board = board;
        temp_board.set_tile(move.first, move.second, colour);

        // Check if we won
        int winner = temp_board.check_winner();
        if ((colour == Colour::RED && winner == 1) ||
            (colour == Colour::BLUE && winner == 2)) {
            return move;
        }
    }

    return {-1, -1};
}

int simulate_with_heuristics(MCTSNode* node, Colour agent_colour) {
    Board board = node->board;
    Colour current_player = node->player_to_move;
    auto empty_tiles = board.get_empty_tiles();

    int move_count = 0;
    static thread_local std::mt19937 gen(std::random_device{}());
    std::uniform_real_distribution<> dis(0.0, 1.0);

    while (!empty_tiles.empty()) {
        // Check for winner periodically
        if (move_count % 5 == 0 || empty_tiles.size() < 10) {
            int winner = board.check_winner();
            if (winner != 0) {
                return winner;
            }
        }

        // Use heuristic-guided selection with 60% probability
        std::pair<int, int> move;
        if (empty_tiles.size() > 5 && dis(gen) < 0.6) {
            move = select_weighted_move(empty_tiles, board, current_player);
        } else {
            // Pure random for speed in late game
            std::uniform_int_distribution<> idx_dist(0, empty_tiles.size() - 1);
            int idx = idx_dist(gen);
            move = empty_tiles[idx];
        }

        // Remove move from empty tiles
        empty_tiles.erase(
            std::remove(empty_tiles.begin(), empty_tiles.end(), move),
            empty_tiles.end()
        );

        // Apply move
        board.set_tile(move.first, move.second, current_player);
        current_player = opposite_colour(current_player);
        move_count++;
    }

    // Final winner check
    return board.check_winner();
}

void backpropagate(MCTSNode* node, int winner, Colour agent_colour,
                   TranspositionTable& tt) {
    while (node != nullptr) {
        node->visits++;

        if (winner != 0) {
            // The node stores player_to_move, so the PARENT made the move
            // We reward the parent if the winner matches who made the move
            if (node->parent != nullptr) {
                Colour parent_player = node->parent->player_to_move;
                if ((parent_player == Colour::RED && winner == 1) ||
                    (parent_player == Colour::BLUE && winner == 2)) {
                    node->wins += 1.0;
                }
            } else {
                // Root node
                if ((agent_colour == Colour::RED && winner == 1) ||
                    (agent_colour == Colour::BLUE && winner == 2)) {
                    node->wins += 1.0;
                }
            }
        }

        // Store updated statistics in transposition table
        if (node->get_hash() != 0) {
            // Compute depth (distance from leaf)
            int depth = 0;
            MCTSNode* temp = node;
            while (temp->parent != nullptr) {
                depth++;
                temp = temp->parent;
            }

            // Find best child move if node has children (for best_move storage)
            std::pair<int, int> best_move = {-1, -1};
            if (!node->children.empty()) {
                auto best_it = std::max_element(
                    node->children.begin(),
                    node->children.end(),
                    [](const auto& a, const auto& b) {
                        return a.second->visits < b.second->visits;
                    }
                );
                best_move = best_it->first;
            }

            // Store in transposition table
            tt.store(node->get_hash(), best_move, node->visits,
                     node->wins, depth, BoundType::EXACT);
        }

        node = node->parent;
    }
}

/**
 * MCTS search with transposition table integration.
 *
 * Transposition Table Features:
 * - Zobrist hashing for position identification
 * - Depth-preferred replacement strategy
 * - Statistical blending during selection (blend_factor=0.7, min_visits=10)
 * - Move ordering during expansion (min_visits=5)
 * - Automatic storage during backpropagation
 *
 * Memory: ~320MB for 5 million entries
 * Expected improvement: 5-40% faster convergence depending on game phase
 */
std::pair<int, int> search(
    const Board& board_state,
    Colour current_player,
    double time_limit
) {
    // Initialize Zobrist hashing
    init_zobrist();

    auto start_time = std::chrono::steady_clock::now();
    double deadline_seconds = time_limit - 0.05; // Reserve 50ms
    auto deadline = start_time + std::chrono::duration<double>(deadline_seconds);

    // Create transposition table for position caching
    TranspositionTable tt(5'000'000);

    // Compute Zobrist hash for root position
    uint64_t root_hash = compute_board_hash(board_state);
    MCTSNode root(board_state, current_player, nullptr, {-1, -1}, root_hash);

    // If only one move available, return it immediately
    if (root.untried_moves.size() == 1) {
        return root.untried_moves[0];
    }

    // Check for immediate winning moves
    auto winning_move = find_winning_move(board_state, current_player);
    if (winning_move.first != -1) {
        return winning_move;
    }

    int iterations = 0;
    int best_move_stable_count = 0;
    std::pair<int, int> last_best_move = {-1, -1};

    while (std::chrono::steady_clock::now() < deadline) {
        MCTSNode* node = &root;

        // Selection: traverse tree using UCB1
        while (node->is_fully_expanded() && !node->children.empty()) {
            node = node->best_child();
        }

        // Check transposition table for this position
        TTEntry cached_entry;
        if (node->get_hash() != 0 && tt.probe(node->get_hash(), cached_entry)) {
            // Only use cached data if it's from sufficient search depth
            if (cached_entry.bound_type == BoundType::EXACT &&
                cached_entry.visits >= 10) {  // Minimum visit threshold

                // Update node statistics with cached values
                // We blend rather than replace to maintain tree consistency
                double blend_factor = 0.7;  // Weight towards cached data
                node->visits = static_cast<int>(
                    node->visits * (1 - blend_factor) +
                    cached_entry.visits * blend_factor
                );
                node->wins = node->wins * (1 - blend_factor) +
                             cached_entry.wins * blend_factor;
            }
        }

        // Expansion: add a new child if not terminal
        if (!node->is_terminal() && !node->is_fully_expanded()) {
            std::pair<int, int> tt_best_move = {-1, -1};
            TTEntry cached_entry;
            if (node->get_hash() != 0 && tt.probe(node->get_hash(), cached_entry)) {
                if (cached_entry.bound_type == BoundType::EXACT &&
                    cached_entry.visits >= 5) {
                    tt_best_move = cached_entry.best_move;
                }
            }
            node = node->expand(tt_best_move);
        }

        // Simulation: heuristic-guided playout
        int winner = simulate_with_heuristics(node, current_player);

        // Backpropagation
        backpropagate(node, winner, current_player, tt);

        iterations++;

        // Early termination check (every 100 iterations)
        if (iterations % 100 == 0 && !root.children.empty()) {
            // Find current best move (most visited)
            auto current_best_it = std::max_element(
                root.children.begin(),
                root.children.end(),
                [](const auto& a, const auto& b) {
                    return a.second->visits < b.second->visits;
                }
            );

            std::pair<int, int> current_best = current_best_it->first;

            if (current_best == last_best_move) {
                best_move_stable_count++;

                // If best move is stable and has high win rate, can terminate early
                if (best_move_stable_count >= 5) {
                    MCTSNode* best_child = current_best_it->second.get();
                    if (best_child->visits > 0) {
                        double win_rate = best_child->wins / best_child->visits;
                        if (win_rate > 0.85 || win_rate < 0.15) {
                            break;
                        }
                    }
                }
            } else {
                best_move_stable_count = 0;
                last_best_move = current_best;
            }
        }
    }

#ifdef TT_DEBUG
    // Transposition table statistics for debugging
    std::cerr << "=== Transposition Table Statistics ===" << std::endl;
    std::cerr << "  Hit Rate: " << (tt.hit_rate() * 100.0) << "%" << std::endl;
    std::cerr << "======================================" << std::endl;
#endif

    // Select best move (most visited child)
    if (!root.children.empty()) {
        auto best_it = std::max_element(
            root.children.begin(),
            root.children.end(),
            [](const auto& a, const auto& b) {
                return a.second->visits < b.second->visits;
            }
        );
        return best_it->first;
    }

    // Fallback: use heuristic-guided selection
    auto empty = board_state.get_empty_tiles();
    if (!empty.empty()) {
        auto move = select_weighted_move(empty, board_state, current_player);
        if (move.first != -1) {
            return move;
        }

        // Random fallback
        static thread_local std::mt19937 gen(std::random_device{}());
        std::uniform_int_distribution<> dis(0, empty.size() - 1);
        return empty[dis(gen)];
    }

    return {0, 0}; // Should never happen
}

} // namespace hex
